falcon/util/streams.py

```diff
@@ -0,0 +1,276 @@
+            delimiter_pos = self._buffer.find(delimiter, self._buffer_pos)
+            if delimiter_pos < 0 and missing_delimiter_error:
+                raise missing_delimiter_error(
+                    'unexpected EOF without delimiter')
+
+        if delimiter_pos >= 0:
+            amount = min(amount, have_bytes + delimiter_pos - self._buffer_pos)
+
+        if have_bytes == 0:
+            # PERF(vytas) Do not join bytes unless needed.
+            ret_value = self._read(amount)
+        else:
+            backlog.append(self._read(amount - have_bytes))
+            ret_value = b''.join(backlog)
+
+        if next_chunk_len > 0:
+            if self._buffer_len == 0:
+                self._buffer = next_chunk
+                self._buffer_len = next_chunk_len
+            else:
+                self._buffer = self._buffer[self._buffer_pos:] + next_chunk
+                self._buffer_len = (self._buffer_len - self._buffer_pos +
+                                    next_chunk_len)
+                self._buffer_pos = 0
+
+        return ret_value
+
+    def _read_until(self, delimiter, amount, missing_delimiter_error=None):
+        # PERF(vytas) In Cython, bind types:
+        #   cdef _read_until(...)
+        #   cdef Py_ssize_t amount
+        #   cdef result
+        #   cdef Py_ssize_t have_bytes = 0
+        #   cdef Py_ssize_t delimiter_len_1
+        #   cdef Py_ssize_t delimiter_pos = -1
+        #   cdef Py_ssize_t offset
+
+        result = []
+        have_bytes = 0
+        delimiter_len_1 = len(delimiter) - 1
+        delimiter_pos = -1
+
+        if not 0 <= delimiter_len_1 < self._chunk_size:
+            raise ValueError('delimiter length must be within [1, chunk_size]')
+
+        while True:
+            if self._buffer_len > self._buffer_pos:
+                delimiter_pos = self._buffer.find(delimiter, self._buffer_pos)
+                if delimiter_pos >= 0:
+                    return self._finalize_read_until(
+                        amount, result, have_bytes,
+                        delimiter_pos=delimiter_pos)
+
+            if amount < (have_bytes + self._buffer_len - self._buffer_pos -
+                         delimiter_len_1):
+                # NOTE(vytas) We now have enough data in the buffer to
+                # return to the caller.
+                return self._finalize_read_until(amount, result, have_bytes)
+
+            read_amount = self._chunk_size
+            if read_amount > self._max_bytes_remaining:
+                read_amount = self._max_bytes_remaining
+            if read_amount == 0:
+                return self._finalize_read_until(
+                    amount, result, have_bytes, delimiter, delimiter_pos,
+                    missing_delimiter_error=missing_delimiter_error)
+
+            next_chunk = self._read_func(read_amount)
+            next_chunk_len = len(next_chunk)
+            self._max_bytes_remaining -= read_amount
+            if next_chunk_len < self._chunk_size:
```

Comment:
I was developing under the assumption that at least WSGI input wrappers must block and return the requested number of bytes, otherwise it is considered an EOF.
But maybe this was a false assumption... I'll reread what the PEPs say.

Also, it might make sense supporting such read functions anyway... I'll check this.